### PR TITLE
Res-1447 knowledge distillation

### PR DESCRIPTION
**Pull request by lucasosouza**
Adds knowledge distillation to ImageNet experiment.

Experiments details in JIRA RES-1447

### CODE REVIEW

File: nupic/research/frameworks/pytorch/model_utils.py [hunk 1]
```diff
@@ -72,6 +74,11 @@ def train_model(
     :param progress_bar: Optional :class:`tqdm` progress bar args.
                          None for no progress bar
     :type progress_bar: dict or None
+    :param teacher model: Teacher model used for knowledge distillation
+    :type teacher model: torch.nn.Module
+    :param kd_factor: Determines the percentage of the target that comes
+                      from the teacher model.
+    :type kd_factor: int or float
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/model_utils.py:81]: How can this be an int?

File: nupic/research/frameworks/pytorch/model_utils.py [hunk 2]
```diff
@@ -111,8 +118,25 @@ def train_model(
...
+                # target is linear combination of teacher and target softmaxes
+                softmax_output_teacher = F.softmax(teacher_model(data))
+                one_hot_target = F.one_hot(target, num_classes=output.shape[-1])
+                combined_target = (kd_factor * softmax_output_teacher
+                                   + (1 - kd_factor) * one_hot_target)
+            # requires a custom loss function.
+            del softmax_output_teacher, one_hot_target
+            loss = soft_cross_entropy(output, combined_target)
+            del combined_target
+        # regular training
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/model_utils.py:136]: This code seems to be very specific for this specific algorithm to be part of this generic training function. 
ImagenetExperiment has an option for custom trainers ([train_model_func](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py#L157-L159)) that can be used for special training cases like this one. 
if you like I can point you to various experiments Michaelangelo created using custom trainers that you can use a reference.

File: nupic/research/frameworks/pytorch/model_utils.py [hunk 3]
```diff
@@ -274,3 +298,41 @@ def deserialize_state_dict(fileobj, device=None):
...
         state_dict = torch.load(fileobj, map_location=device)
     return state_dict
+
+
+def soft_cross_entropy(output, target, size_average=True):
+    """ Cross entropy that accepts soft targets
+    Args:
+         output: predictions for neural network
+         targets: targets, can be soft
+         size_average: if false, sum is returned instead of mean
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/model_utils.py:308]: Avoid mixed docstring styles on the same file. To be consistent the docstring should be:
```
         :param output: predictions for neural network
         :param targets: targets, can be soft
         :param size_average: if false, sum is returned instead of mean
```

File: requirements.txt [hunk 4]
```diff
@@ -46,3 +46,4 @@ tabulate
 torchvision==0.5.0
 tqdm
 h5py
+pretrainedmodels
```
Comment (lscheinkman) [inline requirements.txt:49]: Are you using `se_resnext50_32x4d`? This dependency is only need for this network.
If you really need this dependency i would recommend adding the version number you tested to the requirements.txt. 
This is not an official site for pretrained models and I would be worried about using the latest releases without testing first.

Comment (lucasosouza) [reply requirements.txt:49]: I am using it, if possible I would like to keep the dependency. It is not an official site, but it is a widely used repository for pretrained models, it has more stars in github than torchvision itself. Sure, I can add the specific version number to requirements.

File: nupic/research/frameworks/pytorch/models/pretrained_models.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+"""
+Pretrained models available for knowledge distillation.
+
+For faster loading in worker nodes, use a shared space for head node and workers
+to access pretrained weights by setting the environment variable TORCH_HOME.
+Preload the weights on head node when running for the first time.
+
+Example:
+export TORCH_HOME=/home/ec2-user/nta/results/torch
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/models/pretrained_models.py:30]: If you think it is ok, yes, as long as there are more people using pretrained models besides me. Currently I add an export to my .bash_profile during the setup commands in ray yaml file, so it is fixed in my end.

File: nupic/research/frameworks/pytorch/model_utils.py [hunk 6]
```diff
@@ -111,8 +118,25 @@ def train_model(
...
+
+        # alternative knowledge distillation training
+        if teacher_model is not None:
+            with torch.no_grad():
+                # target is linear combination of teacher and target softmaxes
+                softmax_output_teacher = F.softmax(teacher_model(data))
+                one_hot_target = F.one_hot(target, num_classes=output.shape[-1])
+                combined_target = (kd_factor * softmax_output_teacher
+                                   + (1 - kd_factor) * one_hot_target)
+            # requires a custom loss function.
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/model_utils.py:132]: The argument you only use one loss function. It is incorporated in the next PR, but we will be combining loss functions - regular cross entropy for hard targets, soft cross entropy for the soft targets, and possibly MSELoss for the logits.

File: requirements.txt (hunk 4 above)
Comment (lucasosouza) [reply requirements.txt:49]: I am using. Despite not being official, it is a widely used library, has more stars than github - the large amount of pretrained models will be useful for us. It is also very lightweight.  

There is an open PR in pretrainedmodels to add support to load these models through torchhub, in the next release we might not need to import it directly.

I can add the versioning in requirements

File: nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py [hunk 7]
```diff
@@ -332,6 +345,25 @@ def setup_experiment(self, config):
         # Register post-epoch hooks. To be used as `self.model.apply(post_epoch_hook)`
         self.post_epoch_hooks = config.get("post_epoch_hooks", [])
 
+        # Teacher model and knowledge distillation variables
+        teacher_model_class = config.get("teacher_model_class", None)
+        if teacher_model_class is not None:
```
[the remark this answers is not in the document]
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/imagenet_experiment.py]: I incorporated the mixin approach. Please see if it addresses this concern

File: nupic/research/frameworks/pytorch/imagenet/common_experiments.py [hunk 8]
```diff
@@ -24,13 +24,16 @@
     ImagenetExperiment,
 )
 
-
 class RezeroedKWinnersImagenetExperiment(mixins.RezeroWeights,
                                          mixins.UpdateBoostStrength,
                                          ImagenetExperiment):
     pass
 
+class KnowledgeDistillationImagenetExperiment(mixins.KnowledgeDistillation,
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/imagenet/common_experiments.py:32]: Is this experiment using sparse weights and k-winners?
If it is then you need to add `mixins.RezeroWeights` and  `mixins.UpdateBoostStrength` to the class

Comment (mrcslws) [inline nupic/research/frameworks/pytorch/imagenet/common_experiments.py:32]: You don't really need to create this shared experiment class unless you find it helpful. I only created this file because Luiz pointed out that I was using the same `RezeroedKWinnersImagenetExperiment` class in a couple of places. It is fine for experiment CONFIG.py files to define their own experiment classes.

File: nupic/research/frameworks/pytorch/imagenet/mixins/knowledge_distillation.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+
+import torch
+import torch.nn.functional as F
+from tqdm import tqdm
+
+class KnowledgeDistillation(object):
+    """
+    Sets the network to learn from a teacher model
+    """
+    def __init__(self):
```
Comment (mrcslws) [inline nupic/research/frameworks/pytorch/imagenet/mixins/knowledge_distillation.py:34]: Make sure you update `self.execution_order`. (I'm guessing you were already planning to do this.)

Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/imagenet/mixins/knowledge_distillation.py:30]: Please check the changes to execution_order, as we discussed
